0000-friends_in_high_kindednesses.md

```diff
@@ -0,0 +1,478 @@
+- Feature Name: associated_type_constructors
+- Start Date: 2016-04-29
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow type constructors to be associated with traits. This is an incremental
+step toward a more general feature commonly called "higher-kinded types," which
+is often ranked highly as a requested feature by Rust users. This specific
+feature (associated type constructors) resolves one of the most common use
+cases for higher-kindedness, is a relatively simple extension to the type
+system compared to other forms of higher-kinded polymorphism, and is forward
+compatible with more complex forms of higher-kinded polymorphism that may be
+introduced in the future.
+
+
+# Motivation
+[motivation]: #motivation
+
+Consider the following trait as a representative motivating example:
+
+```rust
+trait StreamingIterator {
+    type Item<'a>;
+    fn next<'a>(&'a mut self) -> Option<Self::Item<'a>>;
+}
+```
+
+This trait is very useful - it allows for a kind of Iterator which yields
+values which have a lifetime tied to the lifetime of the reference passed to
+`next`. A particular obvious use case for this trait would be an iterator over
+a vector which yields overlapping, mutable subslices with each iteration. Using
+the standard `Iterator` interface, such an implementation would be invalid,
+because each slice would be required to exist for as long as the iterator,
+rather than for as long as the borrow initiated by `next`.
+
+This trait cannot be expressed in Rust as it exists today, because it depends
```

Comment:
Nit: you can express it, but it's awfully painful. You need some an extra trait for each associated "higher-kinded type constructor". See for example [the `Graph` trait here](https://github.com/nikomatsakis/nll/blob/master/graph-algorithms/src/lib.rs), which implements an "iterable" pattern in this way.